datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -176,6 +176,10 @@ impl TreeNodeRewriter for TypeCoercionRewriter {
                     negated,
                 )))
             }
+            Expr::Not(expr) => {
+                let expr = not(get_casted_expr_for_bool_op(&expr, &self.schema)?);
+                Ok(expr)
+            }
             Expr::IsTrue(expr) => {
```

Comment:
It might make sense to remove the catch all  from this `match` expression

https://github.com/apache/arrow-datafusion/blob/bc0ba6a724aaaf312b770451718cbce696de6640/datafusion/optimizer/src/analyzer/type_coercion.rs#L413

And explicitly list out all Expr types so that we don't accidentally another Expr type in the future

Is that something you might be willing to do in another PR @guojidan  ?

Reply:
i will do soon